oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/resourceserver/web/BearerTokenAuthenticationEntryPointTests.java

```diff
@@ -0,0 +1,192 @@
+		BearerTokenError error = new BearerTokenError(BearerTokenErrorCodes.INVALID_TOKEN, HttpStatus.UNAUTHORIZED);
+
+		this.authenticationEntryPoint.commence(request, response,
+				new BearerTokenAuthenticationException(error, error.toString()));
+
+		assertThat(response.getStatus()).isEqualTo(401);
+		assertThat(response.getHeader("WWW-Authenticate")).isEqualTo("Bearer error=\"invalid_token\"");
+	}
+
+	@Test
+	public void commence_BearerTokenAuthenticationExceptionInsufficientScope_ShouldSend403AndHeaderWithError()
+			throws IOException {
+		MockHttpServletRequest request = new MockHttpServletRequest();
+		MockHttpServletResponse response = new MockHttpServletResponse();
+		BearerTokenError error = new BearerTokenError(BearerTokenErrorCodes.INSUFFICIENT_SCOPE, HttpStatus.FORBIDDEN);
+
+		this.authenticationEntryPoint.commence(request, response,
+				new BearerTokenAuthenticationException(error, error.toString()));
+
+		assertThat(response.getStatus()).isEqualTo(403);
+		assertThat(response.getHeader("WWW-Authenticate")).isEqualTo("Bearer error=\"insufficient_scope\"");
+	}
+
+	@Test
+	public void commence_BearerTokenAuthenticationExceptionInsufficientScope_ShouldSend403AndHeaderWithErrorAndScope()
+			throws IOException {
+		MockHttpServletRequest request = new MockHttpServletRequest();
+		MockHttpServletResponse response = new MockHttpServletResponse();
+		BearerTokenError error = new BearerTokenError(BearerTokenErrorCodes.INSUFFICIENT_SCOPE, HttpStatus.FORBIDDEN,
+				null, null, "test.read test.write");
+
+		this.authenticationEntryPoint.commence(request, response,
+				new BearerTokenAuthenticationException(error, error.toString()));
+
+		assertThat(response.getStatus()).isEqualTo(403);
+		assertThat(response.getHeader("WWW-Authenticate"))
+				.isEqualTo("Bearer error=\"insufficient_scope\", scope=\"test.read test.write\"");
+	}
+
+	@Test
+	public void commence_BearerTokenAuthenticationExceptionInsufficientScope_ShouldSend403AndHeaderWithErrorAndAllDetails()
+			throws IOException {
+		MockHttpServletRequest request = new MockHttpServletRequest();
+		MockHttpServletResponse response = new MockHttpServletResponse();
+		BearerTokenError error = new BearerTokenError(BearerTokenErrorCodes.INSUFFICIENT_SCOPE, HttpStatus.FORBIDDEN,
+				"Insufficient scope", "http://example.com", "test.read test.write");
+
+		this.authenticationEntryPoint.setRealmName("test");
+		this.authenticationEntryPoint.commence(request, response,
+				new BearerTokenAuthenticationException(error, error.toString()));
+
+		assertThat(response.getStatus()).isEqualTo(403);
+		assertThat(response.getHeader("WWW-Authenticate")).isEqualTo(
+				"Bearer realm=\"test\", error=\"insufficient_scope\", error_description=\"Insufficient scope\", "
+						+ "error_uri=\"http://example.com\", scope=\"test.read test.write\"");
+	}
+
+	@Test
+	public void setRealmName_NullRealmName_ShouldThrowException() {
+		this.thrown.expect(IllegalArgumentException.class);
```

Comment:
Forgot to mention this one; it came up in a conversation I had with @rwinch about how some tests in Spring Security use `ExpectedException` and some use `assertThatThrownBy` (and many older tests use `@Test(expected)`).

We have been moving towards `assertThatThrownBy` since it will make it easier to migrate to JUnit 5 down the road.

The rule, as you have it, is nice and readable, of course. Do you have any heartburn over using AssertJ instead of ExpectedException?

Reply:
I agree `assertThatThrownBy` should be preferred for the same reasons (JUnit 5). I'll address that with the first PR update.